### DISCUSSION TITLE
Custom Stylesheets: Help Needed + Roadmap Suggestions

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#101] [General]
PMLC 4.0.0

PML already offers support for custom CSS, which is a key feature to capitalize for promoting the tool, since custom stylesheets are what empower end users to create documents to their own taste.

I've tried to create alternative stylesheets at the [PML Playground] project, which can be found in the [`stylesheets/`][stylesheets/] sub-project. The scaffolding for creating and testing custom stylesheets using [Sass] is all there, ready to be exploited via Rake automation — it couldn't be easier to work with, just edit the Sass sources and type `rake` to update the project and sample HTML files.

The problem is that I'm not so well versed in CSS, and haven't really caught up with the CSS3/HTML5 revolution, so to speak. In the Playground project, users will find tons of useful links to libraries, documentation to the various CSS "schools of thought", etc.

Right now the PML project really needs the support of a volunteer contributor with experience in CSS design. There are tons of free templates when it comes to websites and blogs, but there is very little covering documentation templates (and they differ quite a lot).

Besides creating some alternative native stylesheet to include with the PMLC package, we also need to start documenting and discussing the design principles that guide PML generated HTML tags, in order to be able to provide CSS guidelines that will work in the future.

Here's a draft of the various topics that might be worth looking into:

1. Documenting PML output HTML tags, so that CSS designers have an official reference to work with.
2. Discuss how HTML tags for native elements might be improved, both in terms of HTML5 semantics and support for custom classes (optional or otherwise).
3. Expert HTML5/CSS3 developers could definitely provide useful insights in this respect, in terms of how to improve the HTML tags and classes in order to provide maximum freedom of customization for end users, as well as supporting accessibility for users with disabilities (high priority).

Since PML will evolve in time, eventually supporting large documents partitioning (e.g. book parts, appendixes, etc.), and because PML might be used in different application domains (documentation, websites, blogs, etc.), investing energy to investigate the best approach to balance all this use cases without supporting one at the expenses of the other is crucial. Furthermore, by taking into account the long-term goals it should be possible to avoid rushed decisions which would result in breaking stylesheets backward compatibility when new document structural feature are added.

From what I gathered so far, the essence of all these CSS schools of thoughts is that good design principles in CSS design were devised to prevent CSS breaking disaster as projects grow, and also to allow the creation of HTML templates that are flexible enough when it comes to user customization of CSS templates.

Stylesheets and HTML templates are tightly interdependent, therefore I believe that the PML documentation for these topics should take into account this and provide guidelines in that direction.

So, is there anyone among the current PML users who has enough knowledge of CSS3 and HTML5 to help us in these tasks? And if not, how can we find and involve someone who might?

> **NOTE** — @pml-lang please create add the `HTML templates` and `CSS` labels to this Discussion, to simplify discovery of related topics in the future, and prevent relevant information being lost amidst the clutter.

[PML Playground]: https://github.com/tajmone/pml-playground/ "PML Playground repository"
[stylesheets/]: https://github.com/tajmone/pml-playground/blob/main/stylesheets/ "PML Playground » stylesheets"
[Sass]: https://sass-lang.com/

**Comment by pml-lang** [comment]
> please create add the HTML templates and CSS labels

Done.
